File: non-constructible-change/non-constructible-change.rb
#   Given an array of positive integers representing the values of coins in your
#   possession, write a function that returns the minimum amount of change (the
#   minimum sum of money) that you <b>cannot</b> create. The given coins can have
#   any positive integer value and aren't necessarily unique (i.e., you can have
#   multiple coins of the same value).

#   For example, if you're given <span>coins = [1, 2, 5]</span>, the minimum
#   amount of change that you can't create is <span>4</span>. If you're given no
#   coins, the minimum amount of change that you can't create is <span>1</span>.

# Sample Input:
#   coins = [5, 7, 1, 1, 2, 3, 22]

# Sample Output:
##   20

def non_constructible_change(coins)
  return 1 if coins.empty?

  sorted_coins = coins.sort
  total = 0

  sorted_coins.each do |coin|
    return total + 1 if coin > total + 1
    total += coin
  end

  total + 1
end

puts non_constructible_change([1, 1, 1, 1, 1])

File: algo-expert-io/non-constructible-change/non-constructible-change.js
function nonConstructibleChange(coins) {
  const totalCoins = coins.reduce((accumulator, currentValue) => {
    return accumulator + currentValue;
  }, 0);

  const sortedCoins = coins.sort((a, b) => a - b);
  let total = 0;
  let answer = 0;
  for (let i = 0; i < sortedCoins.length; i++) {
    const currentCoin = sortedCoins[i];

    if (currentCoin > total + 1) {
      answer = total + 1;
      break;
    }
    total += currentCoin;
  }

  if (answer === 0) {
    return totalCoins + 1;
  }

  return answer;
}

nonConstructibleChange([1, 1, 1, 1, 1]);
